pkgs/characters/test/breaks_test.dart

```diff
@@ -0,0 +1,366 @@
+              visit(stateRegionalOdd);
+            }
+          } else if (newState == stateZWJPictographicLookahead) {
+            if (unreachableStates.remove(newState)) {
+              visit(stateZWJPictographic);
+              visit(stateExtend);
+            }
+          } else if (newState == stateInCLookahead ||
+              newState == stateInCLLookahead) {
+            if (unreachableStates.remove(newState)) {
+              visit(stateExtend);
+            }
+          } else {
+            visit(newState);
+          }
+        }
+        if (unreachableStates.isEmpty) {
+          print("Backward states reachable in $step steps");
+          return;
+        }
+      }
+      expect(unreachableStates, isEmpty, reason: "Should be reachable");
+    });
+
+    test("Backward states distinguishable", () {
+      // Classify states into equivalence categories based on whether they
+      // can be distinguished by *n* transitions. Start with all states
+      // indistinguishable, then create new equivalence classes by splitting
+      // existing equivalence classes by whether they transition differ in
+      // whether to break on any input category, or whether they transition
+      // to states that are distinguishable in the existing equivalence.
+      // Continue until no further equivalence classes are introduced,
+      // the equivalence classes are trivial (one element each),
+      // or (as sanity check) at most `idStateCount` rounds.
+      //
+      // Assume that any lookahead state can be distinguished from any other
+      // state.
+      var states = [for (var i = 0; i < idStateCount; i++) i * scaleState];
+      var eqClasses = [states];
+      var eq = Equivalence(eqClasses);
+
+      for (var r = 0; r <= idStateCount; r++) {
+        var nextEq = Equivalence<int>.distinct(states);
+        // Upper bound.
+        for (var eqClass in eqClasses) {
+          for (var i = 0; i < eqClass.length - 1; i++) {
+            var state1 = eqClass[i];
+            nextPair:
+            for (var j = i + 1; j < eqClass.length; j++) {
+              var state2 = eqClass[j];
+              for (var c = 0; c < categoryCount; c++) {
+                var backState1 = moveBack(state1, c);
+                var backState2 = moveBack(state2, c);
+                if ((backState1 ^ backState2) & maskBreak != 0 ||
+                    backState1 >= idStateLookaheadMin ||
+                    backState2 >= idStateLookaheadMin ||
+                    !eq.eq(backState1 & ~maskBreak, backState2 & ~maskBreak)) {
+                  continue nextPair; // Keep distinguishable.
+                }
+              }
+              nextEq.equate(state1, state2);
+            }
+          }
+        }
+        var nextEqClasses = nextEq.classes;
+        if (nextEqClasses.length == eqClasses.length) break;
+        if (nextEqClasses.length == states.length) {
+          print("Backwards states distinguishable in $r steps");
```

Comment:
This test is a sanity check. It's assumed that all the breaking-tests succeed.
These tests just test that the tables are not too big.

The print is just to give me a little information (I found bugs in the test because the value looked wrong).